xdsl/dialects/experimental/halo.py

```diff
@@ -0,0 +1,341 @@
+from math import prod
+from typing import Sequence
+
+from xdsl.printer import Printer
+from xdsl.parser import BaseParser
+from xdsl.utils.hints import isa
+from xdsl.dialects.experimental import stencil
+from xdsl.ir import Operation, SSAValue, ParametrizedAttribute, Attribute, Dialect
+from xdsl.irdl import OptOpAttr, Annotated, Operand, irdl_op_definition, irdl_attr_definition, ParameterDef
+from xdsl.dialects import builtin
+
+
+@irdl_attr_definition
+class HaloExchangeDecl(ParametrizedAttribute):
+    """
+    This declares a region to be "halo-exchanged".
+    The semantics define that the region specified by offset and size
+    is the *received part*. To get the section that should be sent,
+    use the source_area() method to get the source area.
+
+     - offset gives the coordinates from the origin of the stencil field.
+     - size gives the size of the buffer to be exchanged.
+     - source_offset gives a translation (n-d offset) where the data should be
+       read from that is exchanged with the other node.
+     - neighbor gives the offset in rank to the node this data is to be
+       exchanged with
+
+    Example:
+
+        offset = [4, 0]
+        size   = [10, 1]
+        source_offset = [0, 1]
+        neighbor = -1
+
+    To visualize:
+    0   4         14
+        xxxxxxxxxx    0
+        oooooooooo    1
+
+    Where `x` signifies the area that should be received,
+    and `o` the area that should be read from.
+
+    This data will be exchanged with the node of rank (my_rank -1)
+    """
+    name = "halo.exchange_decl"
+
+    offset_: ParameterDef[builtin.DenseArrayBase]
+    size_: ParameterDef[builtin.DenseArrayBase]
+    source_offset_: ParameterDef[builtin.DenseArrayBase]
+    neighbor_: ParameterDef[builtin.IntegerAttr[builtin.i64]]
+
+    @staticmethod
+    def from_tuples(
```

Comment:
I don't like `from_tuples` given it does not expect tuples specificaly. Maybe something like `from_values`?